lib/index.js

```diff
@@ -222,29 +209,23 @@ function decodeSingle (type, arg) {
     return decodeSingle('uint8', arg).toString() === new BN(1).toString()
   } else if (type === 'string') {
     return new Buffer(decodeSingle('bytes', arg), 'utf8').toString()
-  } else if (type.match(/\w+\[\d+\]/)) {
-    // this part handles fixed-length arrays ([2])
+  } else if (isArray(type)) {
+    // this part handles fixed-length arrays ([2]) and variable length ([]) arrays
     // NOTE: we catch here all calls to arrays, that simplifies the rest
     size = parseTypeArray(type)
-    type = type.slice(0, type.indexOf('['))
-
-    ret = []
-    for (i = 0; i < size; i++) {
-      ret.push(decodeSingle(type, arg.slice(i * 32)))
+    var offset = 0
+    if (size === 'dynamic') {
+      size = decodeSingle('uint256', arg.slice(0, 32)).toNumber()
+      offset = 32
     }
-
-    return ret
-  } else if (type.match(/\w+\[\]/)) {
-    // this part handles variable length ([])
-    // NOTE: we catch here all calls to arrays, that simplifies the rest
-    type = type.slice(0, type.indexOf('['))
-    var count = decodeSingle('uint256', arg.slice(0, 32)).toNumber()
-
     ret = []
-    for (i = 1; i < count + 1; i++) {
-      ret.push(decodeSingle(type, arg.slice(i * 32)))
+    type = type.slice(0, type.lastIndexOf('['))
+    var countOffset = parseTypeArray(type) === 'dynamic' ? 32 : 0
+    for (i = 0; i < size; i++) {
+      var decoded = decodeSingle(type, arg.slice(offset))
```

Comment:
You might have to also provide offset to `decodeSingle` here. Please write a test for `uint[][3]` to check.